lightning/src/offers/invoice.rs

```diff
@@ -0,0 +1,1360 @@
+pub struct Invoice {
+	bytes: Vec<u8>,
+	contents: InvoiceContents,
+	signature: Signature,
+}
+
+/// The contents of an [`Invoice`] for responding to either an [`Offer`] or a [`Refund`].
+///
+/// [`Offer`]: crate::offers::offer::Offer
+/// [`Refund`]: crate::offers::refund::Refund
+enum InvoiceContents {
+	/// Contents for an [`Invoice`] corresponding to an [`Offer`].
+	///
+	/// [`Offer`]: crate::offers::offer::Offer
+	ForOffer {
+		invoice_request: InvoiceRequestContents,
+		fields: InvoiceFields,
+	},
+	/// Contents for an [`Invoice`] corresponding to a [`Refund`].
+	///
+	/// [`Refund`]: crate::offers::refund::Refund
+	ForRefund {
+		refund: RefundContents,
+		fields: InvoiceFields,
+	},
+}
+
+/// Invoice-specific fields for an `invoice` message.
+struct InvoiceFields {
+	payment_paths: Vec<(BlindedPath, BlindedPayInfo)>,
+	created_at: Duration,
+	relative_expiry: Option<Duration>,
+	payment_hash: PaymentHash,
+	amount_msats: u64,
+	fallbacks: Option<Vec<FallbackAddress>>,
+	features: Bolt12InvoiceFeatures,
+	signing_pubkey: PublicKey,
+}
+
+impl Invoice {
+	/// Paths to the recipient originating from publicly reachable nodes, including information
+	/// needed for routing payments across them. Blinded paths provide recipient privacy by
+	/// obfuscating its node id.
+	pub fn payment_paths(&self) -> &[(BlindedPath, BlindedPayInfo)] {
+		&self.contents.fields().payment_paths[..]
+	}
+
+	/// Duration since the Unix epoch when the invoice was created.
+	pub fn created_at(&self) -> Duration {
+		self.contents.fields().created_at
+	}
+
+	/// Duration since [`Invoice::created_at`] when the invoice has expired and therefore should no
+	/// longer be paid.
+	pub fn relative_expiry(&self) -> Duration {
+		self.contents.fields().relative_expiry.unwrap_or(DEFAULT_RELATIVE_EXPIRY)
+	}
+
+	/// Whether the invoice has expired.
+	#[cfg(feature = "std")]
+	pub fn is_expired(&self) -> bool {
+		let absolute_expiry = self.created_at().checked_add(self.relative_expiry());
+		match absolute_expiry {
+			Some(seconds_from_epoch) => match SystemTime::UNIX_EPOCH.elapsed() {
+				Ok(elapsed) => elapsed > seconds_from_epoch,
+				Err(_) => false,
+			},
+			None => false,
+		}
+	}
+
+	/// SHA256 hash of the payment preimage that will be given in return for paying the invoice.
+	pub fn payment_hash(&self) -> PaymentHash {
+		self.contents.fields().payment_hash
+	}
+
+	/// The minimum amount required for a successful payment of the invoice.
+	pub fn amount_msats(&self) -> u64 {
+		self.contents.fields().amount_msats
+	}
+
+	/// Fallback addresses for paying the invoice on-chain, in order of most-preferred to
+	/// least-preferred.
+	pub fn fallbacks(&self) -> Vec<Address> {
+		let network = match self.network() {
+			None => return Vec::new(),
+			Some(network) => network,
+		};
+
+		let to_valid_address = |address: &FallbackAddress| {
+			let program = &address.program;
+			if program.len() < 2 || program.len() > 40 {
```

Comment:
I think this may be redundant with the `is_standard` check? Maybe could add a comment?

Reply:
Added comment.